File: python/tiny_agent/lifecycle.py
from datetime import datetime, timezone
from typing import Callable, Protocol


class LifecycleSink(Protocol):
    def emit(self, event: dict) -> None: ...
    def close(self) -> None: ...


class CallbackSink:
    def __init__(self, emit: Callable[[dict], None]): self._emit = emit
    def emit(self, event: dict) -> None: self._emit(event)
    def close(self) -> None: pass


class ExecutionLifecycle:
    def __init__(self, sinks: list[LifecycleSink] | None = None):
        self.sinks = sinks or []
        self.session_id: str | None = None
        self.operations: dict[str, dict] = {}
        self.models: dict[str, dict] = {}
        self.tools: dict[str, dict] = {}
        self.usage: dict[str, dict] = {}
        self.operation_usage: dict[str, dict] = {}
        self.answers: dict[str, tuple[str, str]] = {}

    def observe(self, event: dict) -> None:
        try: self._observe(event)
        except Exception: pass

    def committed(self, facts: list[dict]) -> None:
        try: self._committed(facts)
        except Exception: pass

    def close(self) -> None:
        now = _timestamp()
        for attempt_id in list(self.models): self._complete_model(attempt_id, now, "effect_unknown")
        for started_id, attempt in list(self.tools.items()):
            self._publish(self._tool_completed(started_id, attempt, now, "effect_unknown"))
            self.tools.pop(started_id, None)
        for sink in self.sinks:
            try: sink.close()
            except Exception: pass

    def _observe(self, event: dict) -> None:
        kind = event.get("type")
        if kind == "session.attached":
            self.session_id = event.get("sessionId")
        elif kind == "recovery.attached":
            if not self.session_id: return
            operation_id = event.get("operationId")
            operation_kind = event.get("operationKind")
            if not isinstance(operation_id, str) or operation_kind not in ("run", "compaction"): return
            self.operations[operation_id] = {"kind": operation_kind, "startedAt": _time(event.get("timestamp")), "recovery": True}
            self.operation_usage[operation_id] = _empty_usage()
            self._publish({**event, "type": "operation.recovered", "sessionId": self.session_id, "recovery": True})
            return
        elif kind == "tool.started":
            if not self.session_id or not isinstance(event.get("toolStartedId"), str): return
            self.tools[event["toolStartedId"]] = {**event, "startedAt": _time(event.get("timestamp"))}
            self._publish({**event, "sessionId": self.session_id})
            return
        self._publish(event)

    def _committed(self, facts: list[dict]) -> None:
        transaction_usage = {}
        for fact in facts:
            usage = _usage(fact.get("usage")) if fact.get("kind") == "usage" else None
            attempt_id = fact.get("attemptId")
            if usage is None or not isinstance(attempt_id, str): continue
            transaction_usage[attempt_id] = usage
            self.usage[attempt_id] = usage
            operation_id = fact.get("operationId")
            if isinstance(operation_id, str): self._add_operation_usage(operation_id, usage)
        for fact in facts: self._fact(fact, transaction_usage)

    def _fact(self, fact: dict, transaction_usage: dict[str, dict]) -> None:
        timestamp = _fact_timestamp(fact)
        if fact.get("kind") == "record" and isinstance(fact.get("record"), dict):
            self._record(fact, fact["record"], timestamp)
            return
        entry = fact.get("entry")
        if fact.get("kind") != "entry" or not isinstance(entry, dict): return
        if entry.get("type") == "message":
            message = entry.get("message", {})
            if message.get("role") == "assistant" and isinstance(entry.get("attemptId"), str):
                attempt_id = entry["attemptId"]
                attempt = self.models.get(attempt_id)
                if attempt and isinstance(fact.get("id"), str) and isinstance(message.get("content"), str):
                    self.answers[fact["id"]] = (attempt["operationId"], message["content"])
                self._complete_model(attempt_id, timestamp, "succeeded", transaction_usage.get(attempt_id))
            elif message.get("role") == "tool" and isinstance(entry.get("toolStartedId"), str):
                self._complete_tool(entry["toolStartedId"], entry, timestamp)
            return
        if entry.get("type") != "compaction" or not isinstance(entry.get("operationId"), str): return
        attempt_id = next((key for key, value in self.models.items() if value["operationId"] == entry["operationId"] and value["kind"] == "compaction"), None)
        if attempt_id: self._complete_model(attempt_id, timestamp, "succeeded", self.usage.get(attempt_id))

    def _record(self, fact: dict, record: dict, timestamp: str) -> None:
        if not self.session_id: return
        kind = record.get("type")
        if kind in ("runStarted", "compactionStarted"):
            operation_id = record.get("operationId")
            if not isinstance(operation_id, str): return
            operation_kind = "run" if kind == "runStarted" else "compaction"
            self.operations[operation_id] = {"kind": operation_kind, "startedAt": _time(timestamp), "recovery": False}
            self.operation_usage[operation_id] = _empty_usage()
            self._publish({"type": "operation.started", "timestamp": timestamp, "sessionId": self.session_id, "operationId": operation_id, "operationKind": operation_kind, "recovery": False})
            return
        if kind == "stepAttempt":
            operation_id, step_id, attempt_id, attempt = (record.get(key) for key in ("operationId", "stepId", "attemptId", "attempt"))
            if not all(isinstance(value, str) for value in (operation_id, step_id, attempt_id)) or not isinstance(attempt, int): return
            operation_kind = "compaction" if record.get("stepKind") == "compaction" else "run"
            recovery = self.operations.get(operation_id, {}).get("recovery", False)
            self.models[attempt_id] = {"operationId": operation_id, "kind": operation_kind, "stepId": step_id, "startedAt": _time(timestamp), "recovery": recovery}
            self._publish({"type": "model.started", "timestamp": timestamp, "sessionId": self.session_id, "operationId": operation_id, "operationKind": operation_kind, "stepId": step_id, "attemptId": attempt_id, "attempt": attempt, "recovery": recovery})
            return
        if kind == "stepFailed":
            operation_id, step_id, attempt_id = (record.get(key) for key in ("operationId", "stepId", "attemptId"))
            if not all(isinstance(value, str) for value in (operation_id, step_id, attempt_id)): return
            error_type = record.get("error", {}).get("code")
            outcome = "cancelled" if error_type == "aborted" else "failed"
            operation = self.operations.get(operation_id)
            if attempt_id not in self.models and operation and operation["recovery"]:
                event = {"type": "model.reconciled", "timestamp": timestamp, "sessionId": self.session_id, "operationId": operation_id, "operationKind": operation["kind"], "stepId": step_id, "attemptId": attempt_id, "recovery": True, "outcome": outcome}
                if isinstance(error_type, str): event["errorType"] = error_type
                self._publish(event)
                return
            self._complete_model(attempt_id, timestamp, outcome, self.usage.get(attempt_id), error_type)
            return
        if kind == "toolStarted":
            values = [fact.get("id"), record.get("operationId"), record.get("stepId"), record.get("toolCallId"), record.get("toolName")]
            if not all(isinstance(value, str) for value in values) or record.get("replay") not in ("safe", "never"): return
            started_id, operation_id, step_id, call_id, tool = values
            self._publish({"type": "tool.admitted", "timestamp": timestamp, "sessionId": self.session_id, "operationId": operation_id, "stepId": step_id, "toolStartedId": started_id, "toolCallId": call_id, "tool": tool, "replay": record["replay"], "recovery": self.operations.get(operation_id, {}).get("recovery", False)})
            return
        if kind == "abortRequested":
            operation_id = record.get("operationId")
            if not isinstance(operation_id, str): return
            event = {"type": "cancel.requested", "timestamp": timestamp, "sessionId": self.session_id, "operationId": operation_id, "operationKind": record.get("operationKind"), "phase": record.get("phase"), "recovery": self.operations.get(operation_id, {}).get("recovery", False)}
            if isinstance(record.get("toolCallId"), str): event["toolCallId"] = record["toolCallId"]
            self._publish(event)
            return
        if kind != "operationFinished" or not isinstance(record.get("operationId"), str): return
        operation_id = record["operationId"]
        operation = self.operations.get(operation_id)
        if not operation: return
        outcome = "succeeded" if record.get("outcome") == "completed" else "cancelled" if record.get("outcome") == "aborted" else "failed"
        event = {"type": "operation.completed", "timestamp": timestamp, "sessionId": self.session_id, "operationId": operation_id, "operationKind": operation["kind"], "recovery": operation["recovery"], "durationMs": _duration(operation["startedAt"], timestamp), "outcome": outcome}
        if record.get("completion") in ("normal", "truncated"): event["completion"] = record["completion"]
        answer = self.answers.get(record.get("finalEntryId"))
        if answer: event["answer"] = answer[1]
        if operation_id in self.operation_usage: event["usage"] = self.operation_usage[operation_id]
        error = record.get("error", {})
        if isinstance(error.get("code"), str): event["errorType"] = error["code"]
        if isinstance(error.get("message"), str): event["errorMessage"] = error["message"]
        self._publish(event)
        self.operations.pop(operation_id, None); self.operation_usage.pop(operation_id, None)
        self.answers = {key: value for key, value in self.answers.items() if value[0] != operation_id}

    def _complete_model(self, attempt_id: str, timestamp: str, outcome: str, usage: dict | None = None, error_type: str | None = None) -> None:
        attempt = self.models.pop(attempt_id, None)
        if not self.session_id or not attempt: return
        event = {"type": "model.completed", "timestamp": timestamp, "sessionId": self.session_id, "operationId": attempt["operationId"], "operationKind": attempt["kind"], "stepId": attempt["stepId"], "attemptId": attempt_id, "recovery": attempt["recovery"], "durationMs": _duration(attempt["startedAt"], timestamp), "outcome": outcome}
        if usage: event["usage"] = _with_cache_rate(usage)
        if isinstance(error_type, str): event["errorType"] = error_type
        self.usage.pop(attempt_id, None); self._publish(event)

    def _complete_tool(self, started_id: str, entry: dict, timestamp: str) -> None:
        attempt = self.tools.pop(started_id, None)
        if not attempt: return
        result = entry.get("result", {})
        outcome = "succeeded" if result.get("type") == "success" else "cancelled" if result.get("reason") == "interrupted" else "failed"
        self._publish(self._tool_completed(started_id, attempt, timestamp, outcome))

    def _tool_completed(self, started_id: str, attempt: dict, timestamp: str, outcome: str) -> dict:
        return {"type": "tool.completed", "timestamp": timestamp, "sessionId": self.session_id, "operationId": attempt["operationId"], "stepId": attempt["stepId"], "attemptId": attempt["attemptId"], "parentAttemptId": attempt["parentAttemptId"], "toolStartedId": started_id, "toolCallId": attempt["toolCallId"], "tool": attempt["tool"], "recovery": attempt["recovery"], "durationMs": _duration(attempt["startedAt"], timestamp), "outcome": outcome}

    def _add_operation_usage(self, operation_id: str, usage: dict) -> None:
        current = self.operation_usage.setdefault(operation_id, _empty_usage())
        for key in _empty_usage(): current[key] += usage[key]
        current["cacheHitRate"] = _with_cache_rate(usage)["cacheHitRate"]

    def _publish(self, event: dict) -> None:
        for sink in self.sinks:
            try: sink.emit(event)
            except Exception: pass


def _empty_usage() -> dict: return {"input": 0, "output": 0, "cacheRead": 0, "cacheWrite": 0}
def _usage(value: object) -> dict | None:
    if not isinstance(value, dict) or any(not isinstance(value.get(key), (int, float)) for key in _empty_usage()): return None
    return {key: value[key] for key in _empty_usage()}
def _with_cache_rate(usage: dict) -> dict:
    prompt = usage["input"] + usage["cacheRead"] + usage["cacheWrite"]
    return {**usage, "cacheHitRate": usage["cacheRead"] / prompt * 100 if prompt else 0}
def _timestamp() -> str: return datetime.now(timezone.utc).isoformat(timespec="milliseconds").replace("+00:00", "Z")
def _fact_timestamp(fact: dict) -> str: return datetime.fromtimestamp(fact.get("timestamp", 0) / 1000, timezone.utc).isoformat(timespec="milliseconds").replace("+00:00", "Z")
def _time(value: object) -> float:
    if not isinstance(value, str): return datetime.now(timezone.utc).timestamp() * 1000
    try: return datetime.fromisoformat(value.replace("Z", "+00:00")).timestamp() * 1000
    except ValueError: return datetime.now(timezone.utc).timestamp() * 1000
def _duration(started: float, timestamp: str) -> float: return max(0, _time(timestamp) - started)

File: typescript/src/lifecycle.ts
import type { SessionFact } from "./session/index.js";

export type Usage = { input: number; output: number; cacheRead: number; cacheWrite: number; cacheHitRate?: number };
export type LifecycleOutcome = "succeeded" | "failed" | "cancelled" | "effect_unknown";

export type LifecycleEvent =
    | {
          type: "startup.started";
          timestamp: string;
          model: string;
          runtime: "typescript";
          plugins: string[];
          mcp: string[];
      }
    | { type: "session.attached"; timestamp: string; sessionId: string; resumed: boolean }
    | { type: "mcp.started"; timestamp: string; server: string }
    | {
          type: "mcp.completed";
          timestamp: string;
          server: string;
          durationMs: number;
          outcome: "succeeded" | "failed";
          protocolVersion?: string;
          toolCount?: number;
          errorType?: string;
      }
    | {
          type: "startup.completed";
          timestamp: string;
          durationMs: number;
          outcome: "succeeded" | "failed";
          errorType?: string;
      }
    | {
          type: "operation.started" | "operation.recovered";
          timestamp: string;
          sessionId: string;
          operationId: string;
          operationKind: "run" | "compaction";
          recovery: boolean;
      }
    | {
          type: "model.started";
          timestamp: string;
          sessionId: string;
          operationId: string;
          operationKind: "run" | "compaction";
          stepId: string;
          attemptId: string;
          attempt: number;
          recovery: boolean;
      }
    | {
          type: "model.completed";
          timestamp: string;
          sessionId: string;
          operationId: string;
          operationKind: "run" | "compaction";
          stepId: string;
          attemptId: string;
          recovery: boolean;
          durationMs: number;
          outcome: "succeeded" | "failed" | "cancelled" | "effect_unknown";
          usage?: Usage;
          errorType?: string;
      }
    | {
          type: "model.reconciled";
          timestamp: string;
          sessionId: string;
          operationId: string;
          operationKind: "run" | "compaction";
          stepId: string;
          attemptId: string;
          recovery: true;
          outcome: "failed" | "cancelled";
          errorType?: string;
      }
    | {
          type: "tool.admitted";
          timestamp: string;
          sessionId: string;
          operationId: string;
          stepId: string;
          toolStartedId: string;
          toolCallId: string;
          tool: string;
          replay: "safe" | "never";
          recovery: boolean;
      }
    | {
          type: "tool.started";
          timestamp: string;
          sessionId: string;
          operationId: string;
          stepId: string;
          attemptId: string;
          parentAttemptId: string;
          toolStartedId: string;
          toolCallId: string;
          tool: string;
          recovery: boolean;
      }
    | {
          type: "tool.completed";
          timestamp: string;
          sessionId: string;
          operationId: string;
          stepId: string;
          attemptId: string;
          parentAttemptId: string;
          toolStartedId: string;
          toolCallId: string;
          tool: string;
          recovery: boolean;
          durationMs: number;
          outcome: "succeeded" | "failed" | "cancelled" | "effect_unknown";
      }
    | {
          type: "cancel.requested";
          timestamp: string;
          sessionId: string;
          operationId: string;
          operationKind: "run" | "compaction";
          phase: "model" | "tool" | "compact";
          toolCallId?: string;
          recovery: boolean;
      }
    | {
          type: "operation.completed";
          timestamp: string;
          sessionId: string;
          operationId: string;
          operationKind: "run" | "compaction";
          recovery: boolean;
          durationMs: number;
          outcome: "succeeded" | "failed" | "cancelled";
          completion?: "normal" | "truncated";
          answer?: string;
          usage?: Usage;
          errorType?: string;
          errorMessage?: string;
      };

export type LifecycleObservation =
    | Extract<
          LifecycleEvent,
          { type: "startup.started" | "session.attached" | "mcp.started" | "mcp.completed" | "startup.completed" }
      >
    | {
          type: "recovery.attached";
          timestamp: string;
          operationId: string;
          operationKind: "run" | "compaction";
      }
    | Omit<Extract<LifecycleEvent, { type: "tool.started" }>, "sessionId">;

export interface LifecycleSink {
    emit(event: LifecycleEvent): void;
    close(): Promise<void>;
}

export interface ExecutionLifecycle {
    observe(event: LifecycleObservation): void;
    committed(facts: readonly SessionFact[]): void;
    close(): Promise<void>;
}

export const noLifecycle: ExecutionLifecycle = {
    observe() {},
    committed() {},
    async close() {},
};

type Operation = {
    operationKind: "run" | "compaction";
    startedAt: number;
    recovery: boolean;
};

type ModelAttempt = {
    operationId: string;
    operationKind: "run" | "compaction";
    stepId: string;
    attemptId: string;
    startedAt: number;
    recovery: boolean;
};

type ToolAttempt = Omit<Extract<LifecycleEvent, { type: "tool.started" }>, "type" | "sessionId" | "timestamp"> & {
    startedAt: number;
};

export class ExecutionLifecycleProjector implements ExecutionLifecycle {
    private sessionId?: string;
    private readonly operations = new Map<string, Operation>();
    private readonly models = new Map<string, ModelAttempt>();
    private readonly tools = new Map<string, ToolAttempt>();
    private readonly usage = new Map<string, Usage>();
    private readonly operationUsage = new Map<string, Usage>();
    private readonly answers = new Map<string, { operationId: string; content: string }>();

    constructor(private readonly sinks: readonly LifecycleSink[]) {}

    observe(event: LifecycleObservation) {
        try {
            this.applyObservation(event);
        } catch {
            // Lifecycle projections are lossy and never change execution.
        }
    }

    committed(facts: readonly SessionFact[]) {
        try {
            this.applyCommitted(facts);
        } catch {
            // A projection bug or sink failure never invalidates a durable commit.
        }
    }

    async close() {
        this.finishPhysicalAttempts();
        this.answers.clear();
        await Promise.all(
            this.sinks.map(async (sink) => {
                try {
                    await sink.close();
                } catch {
                    // Export and flush failures never replace the agent result.
                }
            }),
        );
    }

    private applyObservation(event: LifecycleObservation) {
        if (event.type === "session.attached") this.sessionId = event.sessionId;
        if (event.type === "recovery.attached") {
            const sessionId = this.sessionId;
            if (!sessionId) return;
            const startedAt = timeValue(event.timestamp);
            this.operations.set(event.operationId, {
                operationKind: event.operationKind,
                startedAt,
                recovery: true,
            });
            this.operationUsage.set(event.operationId, emptyUsage());
            this.publish({
                type: "operation.recovered",
                timestamp: event.timestamp,
                sessionId,
                operationId: event.operationId,
                operationKind: event.operationKind,
                recovery: true,
            });
            return;
        }
        if (event.type === "tool.started") {
            const sessionId = this.sessionId;
            if (!sessionId) return;
            const tool = { ...event, startedAt: timeValue(event.timestamp) };
            this.tools.set(event.toolStartedId, tool);
            this.publish({ ...event, sessionId });
            return;
        }
        this.publish(event);
    }

    private applyCommitted(facts: readonly SessionFact[]) {
        const transactionUsage = new Map<string, Usage>();
        for (const fact of facts) {
            if (fact.kind !== "usage" || typeof fact.attemptId !== "string") continue;
            const usage = usageValue(fact.usage);
            if (!usage) continue;
            transactionUsage.set(fact.attemptId, usage);
            this.usage.set(fact.attemptId, usage);
            if (typeof fact.operationId === "string") this.addOperationUsage(fact.operationId, usage);
        }
        for (const fact of facts) this.applyFact(fact, transactionUsage);
    }

    private applyFact(fact: SessionFact, transactionUsage: ReadonlyMap<string, Usage>) {
        const timestamp = timestampValue(fact.timestamp);
        const record = objectValue(fact.record);
        if (fact.kind === "record" && record) {
            this.applyRecord(fact, record, timestamp);
            return;
        }
        const entry = objectValue(fact.entry);
        if (fact.kind !== "entry" || !entry) return;
        if (entry.type === "message") {
            const message = objectValue(entry.message);
            if (message?.role === "assistant" && typeof entry.attemptId === "string") {
                const attempt = this.models.get(entry.attemptId);
                if (attempt && typeof fact.id === "string" && typeof message.content === "string")
                    this.answers.set(fact.id, { operationId: attempt.operationId, content: message.content });
                this.completeModel(entry.attemptId, timestamp, "succeeded", transactionUsage.get(entry.attemptId));
                return;
            }
            if (message?.role === "tool" && typeof entry.toolStartedId === "string") {
                this.completeTool(entry.toolStartedId, entry, timestamp);
            }
            return;
        }
        if (entry.type !== "compaction" || typeof entry.operationId !== "string") return;
        const attempt = [...this.models.values()].find(
            (candidate) => candidate.operationId === entry.operationId && candidate.operationKind === "compaction",
        );
        if (attempt) this.completeModel(attempt.attemptId, timestamp, "succeeded", this.usage.get(attempt.attemptId));
    }

    private applyRecord(fact: SessionFact, record: Record<string, unknown>, timestamp: string) {
        const sessionId = this.sessionId;
        if (!sessionId) return;
        if (record.type === "runStarted" || record.type === "compactionStarted") {
            if (typeof record.operationId !== "string") return;
            const operationKind = record.type === "runStarted" ? "run" : "compaction";
            this.operations.set(record.operationId, {
                operationKind,
                startedAt: timeValue(timestamp),
                recovery: false,
            });
            this.operationUsage.set(record.operationId, emptyUsage());
            this.publish({
                type: "operation.started",
                timestamp,
                sessionId,
                operationId: record.operationId,
                operationKind,
                recovery: false,
            });
            return;
        }
        if (record.type === "stepAttempt") {
            if (
                typeof record.operationId !== "string" ||
                typeof record.stepId !== "string" ||
                typeof record.attemptId !== "string" ||
                typeof record.attempt !== "number"
            )
                return;
            const operation = this.operations.get(record.operationId);
            const operationKind = record.stepKind === "compaction" ? "compaction" : "run";
            const attempt: ModelAttempt = {
                operationId: record.operationId,
                operationKind,
                stepId: record.stepId,
                attemptId: record.attemptId,
                startedAt: timeValue(timestamp),
                recovery: operation?.recovery ?? false,
            };
            this.models.set(record.attemptId, attempt);
            this.publish({
                type: "model.started",
                timestamp,
                sessionId,
                operationId: attempt.operationId,
                operationKind,
                stepId: attempt.stepId,
                attemptId: attempt.attemptId,
                attempt: record.attempt,
                recovery: attempt.recovery,
            });
            return;
        }
        if (record.type === "stepFailed") {
            if (
                typeof record.operationId !== "string" ||
                typeof record.stepId !== "string" ||
                typeof record.attemptId !== "string"
            )
                return;
            const error = objectValue(record.error);
            const cancelled = error?.code === "aborted";
            if (!this.models.has(record.attemptId) && this.operations.get(record.operationId)?.recovery) {
                const operation = this.operations.get(record.operationId);
                if (!operation) return;
                this.publish({
                    type: "model.reconciled",
                    timestamp,
                    sessionId,
                    operationId: record.operationId,
                    operationKind: operation.operationKind,
                    stepId: record.stepId,
                    attemptId: record.attemptId,
                    recovery: true,
                    outcome: cancelled ? "cancelled" : "failed",
                    ...(typeof error?.code === "string" ? { errorType: error.code } : {}),
                });
                return;
            }
            this.completeModel(
                record.attemptId,
                timestamp,
                cancelled ? "cancelled" : "failed",
                this.usage.get(record.attemptId),
                typeof error?.code === "string" ? error.code : undefined,
            );
            return;
        }
        if (record.type === "toolStarted") {
            if (
                typeof fact.id !== "string" ||
                typeof record.operationId !== "string" ||
                typeof record.stepId !== "string" ||
                typeof record.toolCallId !== "string" ||
                typeof record.toolName !== "string" ||
                (record.replay !== "safe" && record.replay !== "never")
            )
                return;
            this.publish({
                type: "tool.admitted",
                timestamp,
                sessionId,
                operationId: record.operationId,
                stepId: record.stepId,
                toolStartedId: fact.id,
                toolCallId: record.toolCallId,
                tool: record.toolName,
                replay: record.replay,
                recovery: this.operations.get(record.operationId)?.recovery ?? false,
            });
            return;
        }
        if (record.type === "abortRequested") {
            if (
                typeof record.operationId !== "string" ||
                (record.operationKind !== "run" && record.operationKind !== "compaction") ||
                (record.phase !== "model" && record.phase !== "tool" && record.phase !== "compact")
            )
                return;
            this.publish({
                type: "cancel.requested",
                timestamp,
                sessionId,
                operationId: record.operationId,
                operationKind: record.operationKind,
                phase: record.phase,
                ...(typeof record.toolCallId === "string" ? { toolCallId: record.toolCallId } : {}),
                recovery: this.operations.get(record.operationId)?.recovery ?? false,
            });
            return;
        }
        if (record.type !== "operationFinished" || typeof record.operationId !== "string") return;
        const operation = this.operations.get(record.operationId);
        if (!operation) return;
        const error = objectValue(record.error);
        const outcome =
            record.outcome === "completed" ? "succeeded" : record.outcome === "aborted" ? "cancelled" : "failed";
        const finalEntryId = typeof record.finalEntryId === "string" ? record.finalEntryId : undefined;
        const usage = this.operationUsage.get(record.operationId);
        this.publish({
            type: "operation.completed",
            timestamp,
            sessionId,
            operationId: record.operationId,
            operationKind: operation.operationKind,
            recovery: operation.recovery,
            durationMs: duration(operation.startedAt, timestamp),
            outcome,
            ...(record.completion === "normal" || record.completion === "truncated"
                ? { completion: record.completion }
                : {}),
            ...(finalEntryId && this.answers.has(finalEntryId)
                ? { answer: this.answers.get(finalEntryId)?.content }
                : {}),
            ...(usage ? { usage } : {}),
            ...(typeof error?.code === "string" ? { errorType: error.code } : {}),
            ...(typeof error?.message === "string" ? { errorMessage: error.message } : {}),
        });
        this.operations.delete(record.operationId);
        this.operationUsage.delete(record.operationId);
        for (const [entryId, answer] of this.answers) {
            if (answer.operationId === record.operationId) this.answers.delete(entryId);
        }
    }

    private completeModel(
        attemptId: string,
        timestamp: string,
        outcome: "succeeded" | "failed" | "cancelled" | "effect_unknown",
        usage?: Usage,
        errorType?: string,
    ) {
        const sessionId = this.sessionId;
        const attempt = this.models.get(attemptId);
        if (!sessionId || !attempt) return;
        this.publish({
            type: "model.completed",
            timestamp,
            sessionId,
            operationId: attempt.operationId,
            operationKind: attempt.operationKind,
            stepId: attempt.stepId,
            attemptId,
            recovery: attempt.recovery,
            durationMs: duration(attempt.startedAt, timestamp),
            outcome,
            ...(usage ? { usage: withCacheHitRate(usage) } : {}),
            ...(errorType ? { errorType } : {}),
        });
        this.models.delete(attemptId);
        this.usage.delete(attemptId);
    }

    private completeTool(toolStartedId: string, entry: Record<string, unknown>, timestamp: string) {
        const sessionId = this.sessionId;
        const attempt = this.tools.get(toolStartedId);
        if (!sessionId || !attempt) return;
        const result = objectValue(entry.result);
        const outcome =
            result?.type === "success" ? "succeeded" : result?.reason === "interrupted" ? "cancelled" : "failed";
        this.publish({
            type: "tool.completed",
            timestamp,
            sessionId,
            operationId: attempt.operationId,
            stepId: attempt.stepId,
            attemptId: attempt.attemptId,
            parentAttemptId: attempt.parentAttemptId,
            toolStartedId,
            toolCallId: attempt.toolCallId,
            tool: attempt.tool,
            recovery: attempt.recovery,
            durationMs: duration(attempt.startedAt, timestamp),
            outcome,
        });
        this.tools.delete(toolStartedId);
    }

    private finishPhysicalAttempts() {
        const sessionId = this.sessionId;
        if (!sessionId) return;
        const timestamp = new Date().toISOString();
        for (const attempt of this.models.values()) {
            this.completeModel(attempt.attemptId, timestamp, "effect_unknown", this.usage.get(attempt.attemptId));
        }
        for (const [toolStartedId, attempt] of this.tools) {
            this.publish({
                type: "tool.completed",
                timestamp,
                sessionId,
                operationId: attempt.operationId,
                stepId: attempt.stepId,
                attemptId: attempt.attemptId,
                parentAttemptId: attempt.parentAttemptId,
                toolStartedId,
                toolCallId: attempt.toolCallId,
                tool: attempt.tool,
                recovery: attempt.recovery,
                durationMs: duration(attempt.startedAt, timestamp),
                outcome: "effect_unknown",
            });
            this.tools.delete(toolStartedId);
        }
    }

    private addOperationUsage(operationId: string, usage: Usage) {
        const current = this.operationUsage.get(operationId) ?? emptyUsage();
        current.input += usage.input;
        current.output += usage.output;
        current.cacheRead += usage.cacheRead;
        current.cacheWrite += usage.cacheWrite;
        current.cacheHitRate = withCacheHitRate(usage).cacheHitRate;
        this.operationUsage.set(operationId, current);
    }

    private publish(event: LifecycleEvent) {
        for (const sink of this.sinks) {
            try {
                sink.emit(event);
            } catch {
                // One broken projection must not block the others.
            }
        }
    }
}

export function callbackSink(emit: (event: LifecycleEvent) => void): LifecycleSink {
    return { emit, async close() {} };
}

function objectValue(value: unknown) {
    return value && typeof value === "object" && !Array.isArray(value) ? (value as Record<string, unknown>) : undefined;
}

function usageValue(value: unknown): Usage | undefined {
    const usage = objectValue(value);
    if (!usage) return undefined;
    const input = numberValue(usage.input);
    const output = numberValue(usage.output);
    const cacheRead = numberValue(usage.cacheRead);
    const cacheWrite = numberValue(usage.cacheWrite);
    if (input === undefined || output === undefined || cacheRead === undefined || cacheWrite === undefined)
        return undefined;
    return { input, output, cacheRead, cacheWrite };
}

function emptyUsage(): Usage {
    return { input: 0, output: 0, cacheRead: 0, cacheWrite: 0 };
}

function withCacheHitRate(usage: Usage): Usage {
    const prompt = usage.input + usage.cacheRead + usage.cacheWrite;
    return { ...usage, cacheHitRate: prompt ? (usage.cacheRead / prompt) * 100 : 0 };
}

function numberValue(value: unknown) {
    return typeof value === "number" && Number.isFinite(value) ? value : undefined;
}

function timestampValue(value: unknown) {
    return new Date(typeof value === "number" ? value : Date.now()).toISOString();
}

function timeValue(value: string) {
    const time = new Date(value).getTime();
    return Number.isFinite(time) ? time : Date.now();
}

function duration(startedAt: number, timestamp: string) {
    return Math.max(0, timeValue(timestamp) - startedAt);
}
